app/controllers/miq_policy_controller/alert_profiles.rb

```diff
@@ -168,80 +168,50 @@ def process_alert_profiles(alert_profiles, task)
-        root_node = TreeNodeBuilder.generic_tree_node(
-          "OBJROOT",
-          @assign[:new][:assign_to].ends_with?("-tags") ? "Tags" : ui_lookup(:tables => @assign[:new][:assign_to]),
-          "100/folder_open.png",
-          "",
-          :cfme_no_click => true,
-          :expand        => true,
-          :hideCheckbox  => true
-        )
-        root_node[:children] = []
-        @objects.sort_by { |o| (o.name.presence || o.description).downcase }.each do |o|
-          if @assign[:new][:assign_to].ends_with?("-tags")
-            icon = "100/tag.png"
-          else
-            if @assign[:new][:assign_to] == "ext_management_system"
-              icon = "svg/vendor-#{o.image_name}.svg"
-            elsif @assign[:new][:assign_to] == "resource_pool"
-              icon = o.vapp ? "100/vapp.png" : "100/resource_pool.png"
-            else
-              icon = "100/#{@assign[:new][:assign_to]}.png"
-            end
-          end
-          node = TreeNodeBuilder.generic_tree_node(
-            o.id,
-            choose_node_identifier(o),
-            icon,
-            "",
-            :cfme_no_click => true,
-            :select        => @assign[:new][:objects].include?(o.id) # Check if tag is assigned
-          )
-          root_node[:children].push(node)
-        end
-        tree = TreeBuilder.convert_bs_tree(root_node).to_json
-      end
+    return nil if alert_profile_get_assign_to_objects_empty?
+    if @assign[:new][:assign_to] == "ems_folder"
+      tree = TreeBuilderBelongsToHac.new(:vat_tree,
+                                         :vat,
+                                         @sb,
+                                         true,
+                                         :edit     => @edit,
+                                         :filters  => @filters,
+                                         :group    => @group,
+                                         :selected => @assign[:new][:objects].collect { |f| "EmsFolder_#{f}" })
+    elsif @assign[:new][:assign_to] == "resource_pool"
+      tree = TreeBuilderBelongsToHac.new(:hac_tree,
+                                         :hac,
+                                         @sb,
+                                         true,
+                                         :edit     => @edit,
+                                         :filters  => @filters,
+                                         :group    => @group,
+                                         :selected => @assign[:new][:objects].collect { |f| "ResourcePool_#{f}" })
+    else
+      tree = TreeBuilderAlertProfileObj.new(:object_tree,
+                                            :object,
+                                            @sb,
+                                            true,
+                                            @assign[:new][:assign_to],
+                                            @assign[:new][:cat],
+                                            @assign[:new][:objects])
```

Comment:
But I just made it hash...code style was angry about large number of function parameters... So what to you want?